Convert to using OPA Constraint Framework

maxsmythe opened the pull request:
This implements the ConstraintTemplate/Constraint functionality for Gatekeeper.

There is a lot more to do including:

- ValidatingRego
- MutatingRego
- Audit
- More useful status fields
- More robust logging/debug capabilities
- Stricter/better schema validation

Signed-off-by: Max Smythe <smythe@google.com>

ritazh commented on pkg/webhook/policy.go:
```diff
@@ -52,8 +35,26 @@ var (
+func AddPolicyWebhook(mgr manager.Manager, opa opa.Client) error {
+	validatingWh, err := builder.NewWebhookBuilder().
+		Validating().
+		Name("mutation.styra.com").
```
```suggestion
		Name(*webhookName).
```

ritazh commented on pkg/webhook/policy.go:
```diff
@@ -84,595 +85,33 @@ func AddPolicyWebhook(mgr manager.Manager) error {
-		w.WriteHeader(405)
-		return
+		glog.Errorf("Error executing query: %s", err)
+		admission.ValidationResponse(false, err.Error())
```
```suggestion
		return admission.ValidationResponse(false, err.Error())
```

maxsmythe replied:
Oh wow, good catch

ritazh commented on pkg/webhook/policy.go:
```diff
@@ -52,8 +35,26 @@ var (
+func AddPolicyWebhook(mgr manager.Manager, opa opa.Client) error {
+	validatingWh, err := builder.NewWebhookBuilder().
+		Validating().
+		Name("validation.gatekeeper.sh").
```
Let's use the `webhookName` flag instead of hardcoding it here. https://github.com/open-policy-agent/gatekeeper/blob/master/pkg/webhook/policy.go#L48 we can default it to `validation.gatekeeper.sh` as you have here.

maxsmythe replied:
Yeah, sorry for clobbering your change twice :/

maxsmythe replied:
re-integrated

tsandall commented on pkg/target/regolib/src.rego:
```diff
@@ -0,0 +1,161 @@
+	count(remaining_groups) < 2
+
+	kinds := {input.review.kind.kind, "*"}
+	selected_kinds := {k | k = kindSelector[r].kinds[_]}
```
@maxsmythe I think you intended `r` to bind to the same value from the comprehension above but not that's going to happen. Here's an alternative implementation of the kind selector matching that gives us the semantics that (I think) we want: https://play.openpolicyagent.org/p/DpckwavJZE. LMKWYT.

tsandall replied:
As a follow-up. It would be good to have some tests in place for the kind selector matching.

maxsmythe replied:
Thank you for the alternative kind selector implementation!

I folded a slightly modified version of your code into the latest patch.

maxsmythe commented:
@tsandall 

Thanks for the review!

I added a comment to the watch manager, but the gist is that changes to the underlying CRD resource version will break the watch, so it's a good idea to make sure the roster stays up-to-date independent of actions from a controller.

Re: other comments:

- Switched to using zap across the board

- Mostly it seems that the global variables are used on initialization to keep track of what things need to get built, which enables dependency injection on the built object via type inspection.  This is the purpose of the AddToManager(). The other global I'm aware of is logging, which is a convenience because passing logs from function to function gets old quick and muddies the API. LMK if I'm missing a use of global variables you're concerned about.

- Thanks! I'm glad the tests were recognized.

- Switched the test targets up

Also, signal boost that I copied your modified kind selector Rego code into the target. LMK if this is not okay to do.

tsandall commented:
> LMK if I'm missing a use of global variables you're concerned about.

No, it was just the initialization you described. It just took me a while to find my way around the code because of the indirection but I think once you spend some time with it, it's not so bad.

The rest of the changes look good to me. Feel free to merge whenever you want.

maxsmythe commented:
Awesome, thanks!
